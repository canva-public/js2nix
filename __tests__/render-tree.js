// @ts-check

const { hostDependencyCycles, render, Package, BrokenReference, Key, Id } = require('../lib/print');
const { createTree } = require('./fixtures/util');

describe('Render nix expression', () => {
  it('a single module', () => {
    const tree = hostDependencyCycles(createTree([
      { name: 'A', modules: [] },
    ]));
    expect(render({ tree })).toMatchSnapshot();
  });

  it('missing sha1', () => {
    const tree = hostDependencyCycles(createTree([
      { name: 'A', modules: [] },
    ]));
    tree[Object.keys(tree).pop()].src = {
      type: 'url',
      url: 'https://registry.yarnpkg.com/A/A.tgz',
    }
    expect(render({ tree })).toMatchSnapshot();
  });

  it('ambiguous url: no extention', () => {
    const tree = hostDependencyCycles(createTree([
      { name: 'A', modules: [] },
    ]));    
    tree[Object.keys(tree).pop()].src = {
      type: 'url',
      url: 'https://registry.yarnpkg.com/A/A',
      sha1: 'sha1',
    }
    expect(render({ tree })).toMatchSnapshot();
  });

  it('hosted modules', () => {
    const tree = hostDependencyCycles(createTree([
      { name: 'A', modules: ['B'] },
      { name: 'B', modules: ['A'] },
    ]));
    expect(render({ tree })).toMatchSnapshot();
  }); 

  it('short references clash', () => {
    /** @type import('../lib/print').DepTree<Package | BrokenReference> */
    const tree = createTree([
      { name: 'A', modules: ['B'] },
      { name: 'B', modules: ['C'] },
      { name: 'C', modules: [] },
    ]);

    tree['A'] = new BrokenReference({
      ref: 'A',
      keys: new Set([
        new Key({ id: new Id({ name: 'A' }), version: '1.1.1'}),
        new Key({ id: new Id({ name: 'A' }), version: '2.1.1'}),
        new Key({ id: new Id({ name: 'A' }), version: '5.0.0'}),
      ]),
    });

    expect(render({ tree })).toMatchSnapshot();
  }); 
});
