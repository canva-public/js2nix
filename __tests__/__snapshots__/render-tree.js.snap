// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render nix expression a single module 1`] = `
"# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.

self: super: {
  \\"A@1.1.1\\" = self.buildNodeModule {
    id = { scope = \\"\\"; name = \\"A\\"; };
    version = \\"1.1.1\\";
    src = self.fetchurl {  
      url = \\"https://registry.yarnpkg.com/A/A.tgz\\"; 
      sha1 = \\"sha1\\"; 
    };
  };
  
  # This information is required for automatic overlays generation logic.
  # It's not possible to inspect/reflect on the attrset that is being built via
  # fixed-point function without getting blocked by infinite recursion.
  # See https://github.com/NixOS/nixpkgs/blob/ee7722a7/lib/fixed-points.nix#L19
  #
  __meta__.aliases = {
  

    # Loopback links, for convenience.
    \\"A@1.1.1\\" = \\"A@1.1.1\\";
  };
}
"
`;

exports[`Render nix expression ambiguous url: no extention 1`] = `
"# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.

self: super: {
  \\"A@1.1.1\\" = self.buildNodeModule {
    id = { scope = \\"\\"; name = \\"A\\"; };
    version = \\"1.1.1\\";
    src = self.fetchurl { 
      name = abort ''
        
        
        Failed to recognise the type of the archive of the \`A@1.1.1\` package served from
        \`https://registry.yarnpkg.com/A/A\`

        Override \`\\"A@1.1.1\\".src.name\` attribute in order to provide a file name extension
        to let Nix recognise the type of the archive and unpack it appropriately. For example:

          self: super: {
            \\"A@1.1.1\\" = super.\\"A@1.1.1\\".override (x: {
              src = x.src.override {
                # Let Nix to recognise the type of archive so it unpacks it appropriately
                name = \\"A.tgz\\";
              };
            });
          }
        
      ''; 
      url = \\"https://registry.yarnpkg.com/A/A\\"; 
      sha1 = \\"sha1\\"; 
    };
  };
  
  # This information is required for automatic overlays generation logic.
  # It's not possible to inspect/reflect on the attrset that is being built via
  # fixed-point function without getting blocked by infinite recursion.
  # See https://github.com/NixOS/nixpkgs/blob/ee7722a7/lib/fixed-points.nix#L19
  #
  __meta__.aliases = {
  

    # Loopback links, for convenience.
    \\"A@1.1.1\\" = \\"A@1.1.1\\";
  };
}
"
`;

exports[`Render nix expression hosted modules 1`] = `
"# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.

self: super: {
  \\"A@1.1.1\\" = self.buildNodeModule {
    id = { scope = \\"\\"; name = \\"A\\"; };
    version = \\"1.1.1\\";
    src = self.fetchurl {  
      url = \\"https://registry.yarnpkg.com/A/A.tgz\\"; 
      sha1 = \\"sha1\\"; 
    };
    modules = [
      self.\\"A@1.1.1+B@2.2.2\\"
    ];
  };
  \\"A@1.1.1+A@1.1.1\\" =
    let id = { scope = \\"\\"; name = \\"A\\"; };
    in { inherit id; host = id; };
  \\"A@1.1.1+B@2.2.2\\" = 
    self.\\"B@2.2.2\\".override {
      host = { scope = \\"\\"; name = \\"A\\"; };
      modules = [
        self.\\"A@1.1.1+A@1.1.1\\"
      ];
    };
  \\"B@2.2.2\\" = self.buildNodeModule {
    id = { scope = \\"\\"; name = \\"B\\"; };
    version = \\"2.2.2\\";
    src = self.fetchurl {  
      url = \\"https://registry.yarnpkg.com/B/B.tgz\\"; 
      sha1 = \\"sha1\\"; 
    };
    modules = [
      self.\\"A@1.1.1\\"
    ];
  };
  
  # This information is required for automatic overlays generation logic.
  # It's not possible to inspect/reflect on the attrset that is being built via
  # fixed-point function without getting blocked by infinite recursion.
  # See https://github.com/NixOS/nixpkgs/blob/ee7722a7/lib/fixed-points.nix#L19
  #
  __meta__.aliases = {
  

    # Loopback links, for convenience.
    \\"A@1.1.1\\" = \\"A@1.1.1\\";
    \\"A@1.1.1+A@1.1.1\\" = \\"A@1.1.1+A@1.1.1\\";
    \\"A@1.1.1+B@2.2.2\\" = \\"A@1.1.1+B@2.2.2\\";
    \\"B@2.2.2\\" = \\"B@2.2.2\\";
  };
}
"
`;

exports[`Render nix expression missing sha1 1`] = `
"# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.

self: super: {
  \\"A@1.1.1\\" = self.buildNodeModule {
    id = { scope = \\"\\"; name = \\"A\\"; };
    version = \\"1.1.1\\";
    src = self.fetchurl {  
      url = \\"https://registry.yarnpkg.com/A/A.tgz\\"; 
      sha256 = abort ''
        
        
        Failed to infer \`sha256\` hash of the \`A@1.1.1\` package source from
        \`https://registry.yarnpkg.com/A/A.tgz\`.

        Override \`\\"A@1.1.1\\".src.sha256\` attribute in order to provide this missing piece to Nix.
        For example:
          
          self: super: {
            \\"A@1.1.1\\" = super.\\"A@1.1.1\\".override (x: {
              # The sha256 value is obtained via
              #   nix-prefetch-url https://registry.yarnpkg.com/A/A.tgz
              src = x.src.override { 
                sha256 = \\"<sha256>\\";
              };
            });
          }

      ''; 
    };
  };
  
  # This information is required for automatic overlays generation logic.
  # It's not possible to inspect/reflect on the attrset that is being built via
  # fixed-point function without getting blocked by infinite recursion.
  # See https://github.com/NixOS/nixpkgs/blob/ee7722a7/lib/fixed-points.nix#L19
  #
  __meta__.aliases = {
  

    # Loopback links, for convenience.
    \\"A@1.1.1\\" = \\"A@1.1.1\\";
  };
}
"
`;

exports[`Render nix expression short references clash 1`] = `
"# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.

self: super: {
  \\"A\\" = abort ''


    Unable to refer to the package via the \`A\` name due to versions' clash.
    Consider referring the package directly via one of these names:

      - \`A@1.1.1\`
      - \`A@2.1.1\`
      - \`A@5.0.0\`


  '';
  \\"A@1.1.1\\" = self.buildNodeModule {
    id = { scope = \\"\\"; name = \\"A\\"; };
    version = \\"1.1.1\\";
    src = self.fetchurl {  
      url = \\"https://registry.yarnpkg.com/A/A.tgz\\"; 
      sha1 = \\"sha1\\"; 
    };
    modules = [
      self.\\"B@2.2.2\\"
    ];
  };
  \\"B@2.2.2\\" = self.buildNodeModule {
    id = { scope = \\"\\"; name = \\"B\\"; };
    version = \\"2.2.2\\";
    src = self.fetchurl {  
      url = \\"https://registry.yarnpkg.com/B/B.tgz\\"; 
      sha1 = \\"sha1\\"; 
    };
    modules = [
      self.\\"C@3.3.3\\"
    ];
  };
  \\"C@3.3.3\\" = self.buildNodeModule {
    id = { scope = \\"\\"; name = \\"C\\"; };
    version = \\"3.3.3\\";
    src = self.fetchurl {  
      url = \\"https://registry.yarnpkg.com/C/C.tgz\\"; 
      sha1 = \\"sha1\\"; 
    };
  };
  
  # This information is required for automatic overlays generation logic.
  # It's not possible to inspect/reflect on the attrset that is being built via
  # fixed-point function without getting blocked by infinite recursion.
  # See https://github.com/NixOS/nixpkgs/blob/ee7722a7/lib/fixed-points.nix#L19
  #
  __meta__.aliases = {
  

    # Loopback links, for convenience.
    \\"A@1.1.1\\" = \\"A@1.1.1\\";
    \\"B@2.2.2\\" = \\"B@2.2.2\\";
    \\"C@3.3.3\\" = \\"C@3.3.3\\";
  };
}
"
`;
